### ISSUE TITLE
PLEASE TEST: Bookmarks Changes - new branch

### ISSUE THREAD

**Issue by IgnorantGuru** [IgnorantGuru/spacefm#492]
Users and developers are invited to help test the 'book' branch of SpaceFM, which includes some changes to bookmarks and will eventually become version 1.0.1.  This branch will be under light development for awhile, but should now be usable and reasonably stable.

**IMPORTANT**:  This branch contains quite a few configuration changes and new code.  Please be sure to create a backup copy of your existing configuration before proceeding:

    cp -a ~/.config/spacefm ~/.config/spacefm-backup

And/or start SpaceFM with a temporary fresh config for testing the defaults:

    rm -rf /tmp/spacefm-config-tmp
    spacefm -c /tmp/spacefm-config-tmp

<del>To install the 'book' branch:</del>  [**UPDATE:** The book branch has been merged into next.  Please see the updated [install instructions](https://github.com/IgnorantGuru/spacefm/issues/492#issuecomment-97464752)]
~~~~
# Download & Extract
mkdir /tmp/spacefm-book && cd /tmp/spacefm-book
wget -O spacefm-book.tar.gz https://github.com/IgnorantGuru/spacefm/tarball/book
tar xzf spacefm-book.tar.gz
cd IgnorantGuru-spacefm-*

# Build & Install
./configure --prefix=/usr  # and optionally: --with-gtk3 see ./configure --help
make -s
sudo make install
sudo update-mime-database /usr/share/mime > /dev/null
sudo update-desktop-database -q
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/hicolor
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/Faenza

# Remove Temporary Files
cd / && rm -rf /tmp/spacefm-book
~~~~

You can copy and paste all of those commands into your terminal as a set. Because this branch is still under some development, it is best to update it every few days by repeating the above commands, and check <a href="https://github.com/IgnorantGuru/spacefm/milestones/book">issues related to the book branch</a> and <a href="https://github.com/IgnorantGuru/spacefm/commits/book">see latest changes</a>.

Or, if you are familiar with git, you can clone the whole SpaceFM repo at git@github.com:IgnorantGuru/spacefm.git **or** https://github.com/IgnorantGuru/spacefm.git

This branch should only be packaged or distributed for experimental/alpha packages.  It still has minor bugs and issues.

**What's New**

This branch takes SpaceFM's bookmarks to the next stage, fully integrating them with the menu system.  New abilities include:
* Design Mode can now be used in the Bookmarks menu to add any kind of custom menu item, including bookmarks, applications, custom commands, submenus, and separators.
* Bookmarks can thus now run commands which allow them to open in specific panels, change view settings via socket commands, etc.
* Bookmarks can now open multiple targets (tab groups)
* The bookmarks pane can now descend into submenus.
* The new Follow Dir option, similar to the old behavior, will select the bookmark matching the current directory, and will also find ones in submenus.  This means that if submenus contain similar bookmarks (such as a 'System' submenu used for system maintenance), entering one of the directories will change the listed bookmarks to the submenu containing similar items.  Thus the bookmarks pane now follows a kind of directory context.
* Each bookmark, other custom menu item, or submenu in the Bookmarks menu and side pane can now have individual icons and shortcut keys assigned (set via Properties).
* Bookmarks can be copied to other menus, exported to plugin files, and imported.
* An option to import GTK's bookmarks is available.

This branch does not represent a complete redesign of the bookmarks pane, more the underlying data.  The bookmarks pane still has limited drag-n-drop functionality (reordering only - you can also now use Cut and Paste to move items between submenus).  For further development, this pane could possibly use a tree (although I'm liking the current single list but opinions are welcome), and more drag-n-drop functionality, but those are very involved development projects.  Redesign of that pane would probably be combined with other future development.  So for now the Bookmarks pane is still somewhat limited, but much more capable than the old one.

Previously, SpaceFM stored its bookmarks in the file *~/.config/spacefm/bookmarks*, which used an extended and mostly-compatible format of GTK's bookmark file (*~/.gtk-bookmarks*).  However, GTK's bookmarks file is just a flat list of paths, so it cannot be used to store SpaceFM's custom menu items or submenus.  When you upgrade to the book branch, your current bookmarks will be imported into the new system.  After that, *~/.config/spacefm/bookmarks* will be ignored and unused.

Before testing, <b>please read through the new <a href="http://ignorantguru.github.io/spacefm/spacefm-manual-next-en.html#gui-book">Bookmarks</a> section of the manual</b>.  It's only a page and a half and will give you the new basic functionality.

<del>There are two unresolved issues I'm aware of:  1) Rarely, the New Tab setting is ignored and a bookmark is opened in a new tab regardless.  2) When editing item properties, switching from Command to Bookmark and back may lose some of the command Options permanently, even if you press Cancel.</del>

If you encounter any other issues or want clarification on expected behavior, please just drop a comment below.  Thanks for testing!

**Comment by VastOne** [comment]
Happily testing... :100:

**Comment by Vladimir-csp** [comment]
I can't build it in Debian, ```make[2]: *** No rule to make target '../src/edit-bookmarks.c', needed by 'spacefm.pot'.  Stop.```

**Comment by IgnorantGuru** [comment]
@Vladimir-csp
Thanks, https://github.com/IgnorantGuru/spacefm/commit/6f1de9ebc1d18e0f4905332fdbc0d488a331849a should fix that, I just hadn't pushed it.  You may need to run configure and make clean.

**Comment by Vladimir-csp** [comment]
Yes it helped!
By the way, I`m constantly getting error about src/ptk/ptk-handler.c that should be placed in po/POTFILES.in

**Comment by Vladimir-csp** [comment]
I`ve added the first application in the list to bookmark pane, 2048-qt. Icon is not shown. This application has an xpm icon located in /usr/share/pixmaps/2048-qt.xpm. Icon is shown in application selection dialog and bookmarks menu, but not in bookmarks pane.

**Comment by prodigalgvs** [comment]
Sorry guys, but I had to revert back to 1.0.0.1.  The book version crashed (simply closed without warnin) so repeatedly that it was virtually unusable.  This seemed to occur after having referenced, or working in, folders with graphics or videos.

**Comment by IgnorantGuru** [comment]
@prodigalgvs
> The book version crashed (simply closed without warnin) so repeatedly that it was virtually unusable.

Thanks for testing, but this isn't a useful report.  See the README - BUILD DEBUG section for how to get a backtrace on a crash and provide a full report.  Also note any actions or particular files that precipitate the crash - eg why are you seeing these crashes and others aren't?  Provide reproducible steps when possible and note any settings (view style, etc) that affect it.

**Comment by prodigalgvs** [comment]
It was not meant to be a bug report.

Gordon VS
gvscheik1@cciwireless.ca

**Comment by IgnorantGuru** [comment]
@Vladimir-csp
> getting error about src/ptk/ptk-handler.c that should be placed in po/POTFILES.in

Thanks, this is fixed in 60ff974

> I`ve added the first application in the list to bookmark pane, 2048-qt. Icon is not shown. This application has an xpm icon located in /usr/share/pixmaps/2048-qt.xpm. Icon is shown in application selection dialog and bookmarks menu, but not in bookmarks pane.

82eaf20 makes application icon loading more consistent and will probably correct that issue.

**Comment by IgnorantGuru** [comment]
Some minor feature changes...  **you can now right-click on the path bar and select New Bookmark** to bookmark an existing file or dir path, or a URL.  (Undocumented: If a key shortcut is assigned to New Bookmark, the shortcut will behave as if you selected New Bookmark from the Bookmarks menu unless the path bar has focus and contains a valid target, in which case the path bar contents will be bookmarked.)

Also, **bookmarked URLs will now show a network icon** (unless a custom icon is set).  (Undocumented: the icon chosen is taken from the Icon|Mounted Network setting of the Devices list.  Most people don't change this so I think reusing that icon is sufficient.  Or if none is set, *gtk-network* is used.)

**Comment by IgnorantGuru** [comment]
**Update:**  All known issues in the book branch have been addressed, and my todo list for this branch is empty.  This branch will probably be merged into the next branch within a few days, so any final testing is appreciated.

**Comment by IgnorantGuru** [comment]
Not related to bookmarks, but I have rewritten the file list Refresh function in b657382 in this branch.  It's now not so kludgy and it will reselect and scroll to the previous cursor position.  Testing requested.  (right-click on file list)|View|Refresh.

**Comment by IgnorantGuru** [comment]
**General question:**  How do you think SpaceFM should handle http URLs by default ootb?  Should it attempt to mount davfs via udevil (current default), or should it open them in the user's web browser?

The davfs handler will be included either way, but I can add an http browser handler above it if that's a better default.

**Comment by VastOne** [comment]
My preference would be to let the browser and http browser handler do the work. As much as I loathe using any type of browser anymore, it is still the best at handling urls, IMO

**Comment by Vladimir-csp** [comment]
On the other hand, SpaceFM is a file manager in first place. If someone opens http:// URL in file manager, it is implied to be a filesystem. And since DAV has no URL scheme for itself, http:// is what left for it. I'd vote for webdav handler by default.

**Comment by IgnorantGuru** [comment]
Well I can make it accept davfs:// and webdav:// as a davfs protocol.  The current behavior is based on Vladimir's argument.  But with the new bookmarks, I was wondering if people would like to bookmark http URLs so they can quickly open related web pages from within SpaceFM.  I may do this myself, so was wondering if it would be a more reasonable default.  Probably not many people using davfs compared to http.  Otoh Vladimir is correct that davfs matches a file manager better.

Another option would be to make a smart handler that checks for davfs2 before using udevil, otherwise falling back to the web browser.  (What would be good file(s) to check for to identify davfs support installed?)  Only loss there would be that people may not realize that SpaceFM can mount davfs.

In either case, this would just be a default - you could always switch the handler priority so that http is handled either way.

**Comment by Vladimir-csp** [comment]
With current book I get segfault on refresh. Presumably if there are more than several images in directory. Also tested with clean config.
```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff522dc8a in strlen () from /lib/x86_64-linux-gnu/libc.so.6
```
Can you reproduce it, or should I build debug packages?

**Comment by IgnorantGuru** [comment]
@Vladimir-csp Yeah I see a crash when I have Large Icons enabled in Detailed view at least - no need for a backtrace just yet - thanks.

**Comment by IgnorantGuru** [comment]
@Vladimir-csp segfault should now be fixed.

**Comment by Xylemon** [comment]
Perhaps I'm missing something, but is there a way to open a new tab from the bookmarks panel without it being a permanent setting? Such as when you click a link in your browser with your mouse3 button, it opens a new tab. A setting like this would be preferable.

**Comment by IgnorantGuru** [comment]
@Xylemon I've been considering this - will look at the mouse3 possibility.  Maybe it can work opposite of the New Tab setting.  Not currently a way to do this.  I was also considering an 'Open' item in the menu that works opposite of New Tab.

**Comment by IgnorantGuru** [comment]
Re http, a513998 adds a new "http & webdav" handler to the top of the protocol handlers list when upgrading from 1.0.0 (or from the current next or book branches).

This will mount webdav:// and davfs:// URLs in udevil as http.  It will also mount http:// and https:// URLs in udevil if udevil.conf's allowed_types contains "davfs".

There are some hints at the top of the Mount command for controlling it's behavior.

I figure this is a reasonable default because if the user doesn't have udevil installed with davfs allowed (which must be manually enabled in udevil.conf), the user would just get an error anyway.  So instead it uses the web browser.  So novice users get reasonable behavior for http, and more experienced users can figure out what they want.

One nice thing about this feature is that the Bookmarks follow the dir, so bookmark submenus can contain bookmarks to dirs as well as related websites.

Comments welcome.

**Comment by IgnorantGuru** [comment]
@Xylemon In c81b663, an Open item was added to the Bookmarks side pane context menu which will obey the reverse of the New Tab setting.  A middle-click on a bookmark also reverses it.  Thanks for any testing.

**Comment by IgnorantGuru** [comment]
The book branch has now been merged into next, and will be deleted shortly.  Please update your scripts to use these install instructions for the next branch:

~~~~
# Download & Extract
mkdir /tmp/spacefm-next && cd /tmp/spacefm-next
wget -O spacefm-next.tar.gz https://github.com/IgnorantGuru/spacefm/tarball/next
tar xzf spacefm-next.tar.gz
cd IgnorantGuru-spacefm-*

# Build & Install
./configure --prefix=/usr  # and optionally: --with-gtk3 see ./configure --help
make -s
sudo make install
sudo update-mime-database /usr/share/mime > /dev/null
sudo update-desktop-database -q
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/hicolor
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/Faenza

# Remove Temporary Files
cd / && rm -rf /tmp/spacefm-next
~~~~

The next branch represents a beta version pending release for 1.0.1, so any further testing is appreciated.  Thanks!

**Comment by Xylemon** [comment]
@IgnorantGuru that last commit was perfect! Exactly what I was wanting, thank you! Been great testing this branch.

**Comment by IgnorantGuru** [comment]
This branch has been released in 1.0.1.  Thanks for testing!
